src/config.c

```diff
@@ -89,156 +96,129 @@ int kBudSSLClientIndex = -1;
+  { "version", 0, NULL, 'v' },
+  { "config", 1, NULL, 'c' },
+  { "piped-config", 0, NULL, 'p' },
+  { "inline-config", 1, NULL, 'i' },
+#ifndef _WIN32
+  { "daemonize", 0, NULL, 'd' },
+#endif  /* !_WIN32 */
+  { "worker", 0, NULL, 1000 },
+  { "default-config", 0, NULL, 1001 },
+  { NULL, 0, NULL, 0 }
+};
 
 
-bud_config_t* bud_config_cli_load(int argc, char** argv, bud_error_t* err) {
+bud_error_t bud_config_new(int argc, char** argv, bud_config_t** out) {
+  bud_error_t err;
+  bud_config_t* config;
+  int i;
+  int r;
+  size_t path_len;
   int c;
   int index;
-  int is_daemon;
-  int is_worker;
-  size_t path_len;
-  bud_config_t* config;
-
-  struct option long_options[] = {
-    { "version", 0, NULL, 'v' },
-    { "config", 1, NULL, 'c' },
-    { "piped-config", 0, NULL, 'p' },
-    { "inline-config", 1, NULL, 'i' },
-#ifndef _WIN32
-    { "daemonize", 0, NULL, 'd' },
-#endif  /* !_WIN32 */
-    { "worker", 0, NULL, 1000 },
-    { "default-config", 0, NULL, 1001 },
-    { NULL, 0, NULL, 0 }
-  };
 
-  *err = bud_ok();
-  config = NULL;
-  is_daemon = 0;
-  is_worker = 0;
+  config = calloc(1, sizeof(*config));
+  if (config == NULL)
+    return bud_error_str(kBudErrNoMem, "bud_config_t");
 
   do {
     index = 0;
-    c = getopt_long(argc, argv, "vi:c:dp", long_options, &index);
+    c = getopt_long(argc, argv, bud_long_flags, bud_long_options, &index);
     switch (c) {
       case 'v':
         bud_print_version();
-        c = -1;
+        goto no_config;
+#ifndef _WIN32
+      case 'd':
+        config->is_daemon = 1;
+#endif  /* !_WIN32 */
         break;
       case 'p':
       case 'i':
       case 'c':
-        if (config != NULL) {
-          bud_config_free(config);
-          config = NULL;
-          *err = bud_error(kBudErrMultipleConfigs);
-          return NULL;
+        if (config->path != NULL || config->inlined || config->piped) {
+          err = bud_error(kBudErrMultipleConfigs);
+          goto fatal;
         }
-
-        config = bud_config_load(c == 'p' ? NULL : optarg, c == 'i', err);
-        if (config == NULL) {
-          ASSERT(!bud_is_ok(*err), "Config load failed without error");
-          c = -1;
-          break;
+        if (c == 'p') {
+          config->piped = 1;
+          config->path = "!config";
+        } else {
+          config->piped = 0;
+          config->path = optarg;
+          config->inlined = c == 'i';
         }
-
-        if (is_daemon)
-          config->is_daemon = 1;
-        if (is_worker)
-          config->is_worker = 1;
-        break;
-#ifndef _WIN32
-      case 'd':
-        is_daemon = 1;
-        if (config != NULL)
-          config->is_daemon = 1;
-#endif  /* !_WIN32 */
         break;
       case 1000:
-        is_worker = 1;
-        if (config != NULL)
-          config->is_worker = 1;
+        config->is_worker = 1;
         break;
       case 1001:
         bud_config_print_default();
-        c = -1;
-        break;
+        goto no_config;
```

Comment:
A request to print out the default config should not produce an error complaining about no configuration having been passed in.